lua/whichkey_setup.lua

```diff
@@ -73,41 +92,62 @@ local function handle_user_opts(opts)
             opts[key] = value
         end
     end
-    return opts
+    -- split general options and keymap options
+    local split_opts = {keymap_opts = {}}
+    for key, value in pairs(opts) do
+        if key == "silent" or key == "noremap" then
+            split_opts.keymap_opts[key] = value
+        else
+            split_opts[key] = value
+        end
+    end
+    return split_opts
 end
 
-M.register_keymap = function(leader_type, keymap, opts)
+M.register_keymap = function(initial_key, keymap, opts)
     opts = handle_user_opts(opts)
-    local bufnr = opts.bufnr
-    opts.bufnr = nil
-    local mode, key, mapped_key
-    if (leader_type == 'leader') then
-        mode = 'n'
-        key = '<Leader>'
-        mapped_key = vim.g.mapleader
-    elseif (leader_type == 'localleader') then
-        mode = 'n'
-        key = '<Localleader>'
-        mapped_key = vim.g.maplocalleader
-    elseif (leader_type == 'visual') then
+    -- check if initial_key specifies mode for backwards compatibility
+    local mode = opts.mode
+    if initial_key == 'visual' then
         mode = 'v'
-        key = '<Leader>'
-        mapped_key = '<VisualBindings>'
-    elseif (leader_type == 'localvisual') then
+        initial_key = 'leader'
+    end
+    if initial_key == 'localvisual' then
         mode = 'v'
+        initial_key = 'localleader'
+    end
+    local key, raw_key
+    if (initial_key == 'leader') then -- handle special initial keys
+        key = '<Leader>'
+        if mode == 'n' then
+            raw_key = vim.g.mapleader
+        else
+            raw_key = '<VisualBindings>'
+        end
+    elseif (initial_key == 'localleader') then
         key = '<Localleader>'
-        mapped_key = '<LocalVisualBindings>'
+        if mode == 'n' then
+            raw_key = vim.g.maplocalleader
+        else
+            raw_key = '<LocalVisualBindings>'
+        end
     else
-        vim.cmd('echoerr "Unknown map type '..leader_type..'"')
+        if string.len(initial_key) > 1 then
+            vim.cmd('echoerr "Can only specify a single initial key, not '..initial_key..'"')
+        end
+        key = initial_key
+        raw_key = initial_key
     end
-    local textmap = textmaps[leader_type]
-    setup_keymap(mode, {key}, keymap, textmap, opts, bufnr)
+    local textmap_key = initial_key..'_'..mode
```

Comment:
we probably should add in `bufnr` here is well if it's defined in `opts`. Will be needed once the upstream PR with buffer local which-keymaps is merged.

Reply:
PR: https://github.com/liuchengxu/vim-which-key/pull/48

Reply:
Good point @folke! I guess in most cases people would set `bufnr = 0` to use the current buffer. But I'm wondering how to handle other cases. Since if we store the buffer number in `textmaps` then what if the buffers change? Or do buffer numbers persist in a session? Any thoughts?